src/modm/platform/comp/stm32/comp.hpp.in

```diff
@@ -19,696 +17,435 @@
-			};
-
-			enum BlankingSource
-			{
-				NO_BLANKING = 0b000,
-			{% if id <= 2 -%}
-				TIMER_1_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0,	// 0b001
-				TIMER_2_OC3 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1,	// 0b010
-				TIMER_3_OC3 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1 |
-								COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0	// 0b011
-			{% elif id == 3 -%}
-				TIMER_1_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0,	// 0b001
-				TIMER_2_OC4 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1 |
-								COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0	// 0b011
-			{% elif id == 4 -%}
-				TIMER_3_OC4 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0,	// 0b001
-				TIMER_8_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1,	// 0b010
-				TIMER_15_OC1 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1 |
-								COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0	// 0b011
-			{% elif id == 5 -%}
-				TIMER_8_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1,	// 0b010
-				TIMER_3_OC3 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1 |
-								COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0	// 0b011
-			{% elif id == 6 -%}
-				TIMER_8_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1,	// 0b010
-				TIMER_2_OC4 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1 |
-								COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0,// 0b011
-				TIMER_15_OC2 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_2	// 0b100
-			{% elif id == 7 -%}
-				TIMER_1_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_0,	// 0b001
-				TIMER_8_OC5 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_1,	// 0b010
-				TIMER_15_OC2 = COMP{{ id }}_CSR_COMP{{ id }}BLANKING_2	// 0b100
-			{% endif -%}
-			};
-
-		public:
-
-			/**
-			 * Initialize and enable the comparator.
-			 *
-			 * Enables the comperator and sets important values.
-			 *
-			 * Do NOT set lock = true if you want to be able to set other values
-			 * later.
-			 */
-			static inline void
-			initialize(
-						InvertingInput n_in,
-					{% if id > 1 -%}
-						NonInvertingInput p_in,
-					{% endif -%}
-						Output out = Output::NOT_CONNECTED,
-						Hysteresis hyst = Hysteresis::NO_HYSTERESIS,
-						Mode mode = Mode::HIGH_SPEED,
-						Polarity pol = Polarity::NOT_INVERTED,
-						bool lock_comp = false)
-			{
-				setInvertingInput(n_in);
-			{% if id > 1 -%}
-				setNonInvertingInput(p_in);
 			{% endif -%}
-				setOutputSelection(out);
-				setHysteresis(hyst);
-				setMode(mode);
-				setPolarity(pol);
-				setEnabled(true);	// enable comparator
-				if(lock_comp) lock();
-			}
-
+		};
+	protected:
+		{% if target.family == "f3" -%}
+		static constexpr uint32_t NonInvertingInputMask = 0b1 << 7;
+		{% elif target.family == "l4" -%}
+		static constexpr uint32_t NonInvertingInputMask = 0b11 << 7;
+		{% endif -%}
 
+	{% if target.family == "f3" -%}
+	public:
+		enum class
+		Output
+		{
+			// same in every comparator instance:
+			NoSelection		= 0b0000 << 10,
+			Tim1BkIn		= 0b0001 << 10,
+			Tim1BkIn2		= 0b0010 << 10,
+			Tim8BkIn		= 0b0011 << 10,
+			Tim8BkIn2		= 0b0100 << 10,
+			Tim1Or8BkIn2	= 0b0101 << 10,
```

Comment:
Are… are these the table you need (from F3 refman)? zomg… 🙈 
![](https://user-images.githubusercontent.com/163066/42738759-830636da-8889-11e8-9b87-2d2a9478089b.png)
![](https://user-images.githubusercontent.com/163066/42738775-da084efa-8889-11e8-91c0-277cdc3c14ef.png)

Reply:
Yeah, these tables and/or the register descriptions...

@chris-durand is working on parsing the register description using *pdfminer*...